src/serviceworker/index.ts

```diff
@@ -0,0 +1,161 @@
+    // For example, `/_matrix/media/unstable` or `/_matrix/media/v3/preview_url` are something well within
+    // the control of the application, and appear to be choices made at a higher level than us.
+    if (url.includes("/_matrix/media/v3/download") || url.includes("/_matrix/media/v3/thumbnail")) {
+        // We need to call respondWith synchronously, otherwise we may never execute properly. This means
+        // later on we need to proxy the request through if it turns out the server doesn't support authentication.
+        event.respondWith(
+            (async (): Promise<Response> => {
+                let fetchConfig: { headers?: { [key: string]: string } } = {};
+                try {
+                    // Figure out which homeserver we're communicating with
+                    const csApi = url.substring(0, url.indexOf("/_matrix/media/v3"));
+
+                    // Add jitter to reduce request spam, particularly to `/versions` on initial page load
+                    await new Promise<void>((resolve) => setTimeout(() => resolve(), Math.random() * 10));
+
+                    // Locate our access token, and populate the fetchConfig with the authentication header.
+                    // @ts-expect-error - service worker types are not available. See 'fetch' event handler.
+                    const client = await self.clients.get(event.clientId);
+                    const accessToken = await getAccessToken(client);
+                    if (accessToken) {
+                        fetchConfig = {
+                            headers: {
+                                Authorization: `Bearer ${accessToken}`,
+                            },
+                        };
+                    }
+
+                    // Update or populate the server support map using a (usually) authenticated `/versions` call.
+                    if (!serverSupportMap[csApi] || serverSupportMap[csApi].cacheExpires <= new Date().getTime()) {
+                        const versions = await (await fetch(`${csApi}/_matrix/client/versions`, fetchConfig)).json();
+                        serverSupportMap[csApi] = {
+                            supportsMSC3916: Boolean(versions?.unstable_features?.["org.matrix.msc3916"]),
+                            cacheExpires: new Date().getTime() + 2 * 60 * 60 * 1000, // 2 hours from now
+                        };
+                    }
+
+                    // If we have server support (and a means of authentication), rewrite the URL to use MSC3916 endpoints.
+                    if (serverSupportMap[csApi].supportsMSC3916 && accessToken) {
+                        // Currently unstable only.
+                        // TODO: Support stable endpoints when available.
+                        url = url.replace(/\/media\/v3\/(.*)\//, "/client/unstable/org.matrix.msc3916/media/$1/");
+                    } // else by default we make no changes
```

Comment:
rewriting the URL in the depths of the serviceworker feels very, very magical. Shouldn't this happen in the js-sdk?

Reply:
I was thinking of a method on MatrixClient or something for turning mxc URIs into http URLs. Cache the versions check early on, I guess.

> Instead, the idea is we let the service worker rewrite the url and if that fails or is unsupported, the natural fallback behaviour exists instead.

Seems like we can't rely on this anyway, because we're going to need authentication. Or rather: if it fails often enough that we need that fallback, then the whole approach seems doomed.

Reply:
tldr - js-sdk PR for documentation is here: https://github.com/matrix-org/matrix-js-sdk/pull/4185

> Well: we can't use the authenticated endpoint until we get hold of an access token, and getting hold of an access token is an async operation. That seems to me a good time to decide whether we ought to use that access token in media requests.
> 
> I guess that's a bunch more work, though :(.

The issue is that the call stack is *not* capable of having *any* async code in it. The `getHttpUriForMxc` function in the js-sdk (and `client.mxcUrlToHttp`, which just calls `getHttpUriForMxc`) is *not* async - introducing a promise of any kind would be a breaking change. We could add secondary functions (`getAuthenticatedHttpUriForMxc`, for example), though then we reach the react-sdk layer where the `Media` class exists and is engrained throughout the entire layer, and of course doesn't support async operations in its general cases either. This same problem propagates outwards to the remaining components, functions, classes, etc of the react-sdk. 

It is indeed a bunch more work to fix this, and https://github.com/matrix-org/matrix-react-sdk/commit/5cdd706ef9e34ffc8a31a820c2c9fedf1694a767 only covers some of it (as an alternative I wrote to avoid using service workers). Even if we replace the URL rewrite with a js-sdk function call, we're still collecting blobs and filling up our memory buffers. The hardest things to fix are the hidden uses of non-image elements, like in the case of the composer where avatars are [applied using CSS](https://github.com/matrix-org/matrix-react-sdk/blob/a0c8575113ac74119cd7841e956cc13463ff9244/src/editor/parts.ts#L308). 

Service workers cover all of these nuanced cases, though ideally I think the js-sdk would indeed be producing authentication-first URLs (which the service worker can detect and append an `Authorization` header to). This is however a pretty breaking change to the app, so at least while things are unstable I think it's best to keep the fallback.


> The fact we might be able to get rid of this code in future does little to quell my concerns. Experience strongly suggests that this code will still be here in 8 years time.

This is my experience too, fwiw. I'm hopeful our [rollout plan](https://docs.google.com/document/d/1TgsktJ9ucYDoU6sgkYvcMU9OvkfasMUIEYPpjnT8N7g/edit) (internal only) will help actually make this code temporary and work to eradicate the unauthenticated option, though the service worker itself will need to be maintained. We can at least remove/reduce some of the magic as the MSC progresses towards total stability.


> * We're basically saying the js-sdk is useless without this service worker. That's actually inherent in the fact we have to add authorization, but the less the service worker has to do the better. Anyway, at the very least we need to add documentation to the js-sdk so that people using it understand what their service worker has to do.

I'm not sure I'd classify it as "useless" given servers aren't likely to turn on an unstable feature just yet, but it is something worth documenting. I've done so here: https://github.com/matrix-org/matrix-js-sdk/pull/4185

I suspect it's possible for the js-sdk to support these sorts of operations when it's running in a NodeJS environment, though in a browser environment it's almost certainly a service worker problem given the consumer's (react-sdk) usage.


> * I feel like rewriting URLs in a service worker is subtle, unexpected behaviour. A future generation of developers is going to be extremely confused by this. For reasons I can't entirely explain, it feels worse than "just" adding an Authorization header. Hence, I'm pretty keen we do our best to find alternatives.

Mentioned above, it is pretty magical. With the browser devtools making it clear what's happening and the rollout strategy though, I think we'll be okay. At the very least we can switch to using stable endpoints in the js-sdk pretty quickly by default (once the spec process allows us to), and convert the service worker to 'just' adding the Authorization header in place. This would break compatibility with older servers, but at least on T&S we consider that to be a feature (once the MSC is stable and/or released - doing so *right now* would require servers to advertise unstable functionality, which effectively bypasses the MSC process and leads to defacto spec).

Reply:
> tldr - js-sdk PR for documentation is here: https://github.com/matrix-org/matrix-js-sdk/pull/4185

Thanks for starting to document this. It certainly helps quell some of my concerns here.

> The issue is that the call stack is _not_ capable of having _any_ async code in it. The `getHttpUriForMxc` function in the js-sdk (and `client.mxcUrlToHttp`, which just calls `getHttpUriForMxc`) is _not_ async - introducing a promise of any kind would be a breaking change. 

I don't really follow this. Why can't we make the `/versions` request before we construct a `MatrixClient`, just as we currently make a `/login` request? That would mean that `client.mxcUrlToHttp` could still be synchronous, but do the right thing.

So is the problem that there are some places that we construct media links where there is no access to a `MatrixClient`? Or before a `MatrixClient` is even constructed? It feels like such things probably need to work without authentication anyway?

... which is not to say it's not a bunch of work, but I'm still keen to understand if we're doing this for pragmatic reasons or because it's the only way it can possibly work.

> At the very least we can switch to using stable endpoints in the js-sdk pretty quickly by default (once the spec process allows us to), and convert the service worker to 'just' adding the Authorization header in place.

Well, I'll believe it when I see it...

Reply:
*moving comment into thread*

@t3chguy [says](https://github.com/element-hq/element-web/pull/27326#issuecomment-2088257761):

> > I don't really follow this. Why can't we make the /versions request before we construct a MatrixClient, just as we currently make a /login request? That would mean that client.mxcUrlToHttp could still be synchronous, but do the right thing.
> 
> This would break the offline mode in Element Web

Reply:
> I don't really follow this. Why can't we make the `/versions` request before we construct a `MatrixClient`, just as we currently make a `/login` request? That would mean that `client.mxcUrlToHttp` could still be synchronous, but do the right thing.
> 
> So is the problem that there are some places that we construct media links where there is no access to a `MatrixClient`? Or before a `MatrixClient` is even constructed? It feels like such things probably need to work without authentication anyway?
> 
> ... which is not to say it's not a bunch of work, but I'm still keen to understand if we're doing this for pragmatic reasons or because it's the only way it can possibly work.

`/login` I think is trivial to support. If a user has an existing session though, we seem to show their avatar *very* early in the cycle, making some code a bit racy. The earliest we can check `/versions` is *sometimes* too late, at least as far as my browser is concerned: https://github.com/matrix-org/matrix-react-sdk/commit/0a1d45c0f91e9567211affe45670ac5ae1a6d502 (this commit has other problems, like lack of re-checking server support, linting, etc).

The service worker avoids the lack of `MatrixClient` by reaching into idb directly for the token. The `postMessage` for user information also reaches directly into localstorage, avoiding `MatrixClient` too.

This is out of pragmatism imo, though I'm still open to suggestions/hints on where to put this code otherwise.